src/OpenTelemetry.Contrib.Instrumentation.EventCounters/Implementation/EventCounterListener.cs

```diff
@@ -0,0 +1,243 @@
+                }
+
+                EventCountersInstrumentationEventSource.Log.EventCounterInitializeSuccess();
+                this.isInitialized = true;
+
+                // Go over every EventSource created before we finished initialization, and enable if required.
+                // This will take care of all EventSources created before initialization was done.
+                foreach (var eventSource in this.allEventSourcesCreated)
+                {
+                    this.EnableEventSource(eventSource);
+                }
+            }
+            catch (Exception ex)
+            {
+                EventCountersInstrumentationEventSource.Log.EventCountersInstrumentationError("EventCounterListener Constructor", ex.Message);
+            }
+        }
+
+        /// <summary>
+        /// Processes a new EventSource event.
+        /// </summary>
+        /// <param name="eventData">Event to process.</param>
+        protected override void OnEventWritten(EventWrittenEventArgs eventData)
+        {
+            // Ignore events if initialization not done yet. We may lose the 1st event if it happens before initialization, in multi-thread situations.
+            // Since these are counters, losing the 1st event will not have noticeable impact.
+            if (!this.isInitialized)
+            {
+                return;
+            }
+
+            try
+            {
+                if (this.countersToCollect.ContainsKey(eventData.EventSource.Name))
+                {
+                    if (eventData.Payload.Count > 0 && eventData.Payload[0] is IDictionary<string, object> eventPayload)
+                    {
+                        this.ExtractAndPostMetric(eventData.EventSource.Name, eventPayload);
+                    }
+                    else
+                    {
+                        EventCountersInstrumentationEventSource.Log.IgnoreEventWrittenAsEventPayloadNotParseable(eventData.EventSource.Name);
+                    }
+                }
+                else
+                {
+                    EventCountersInstrumentationEventSource.Log.IgnoreEventWrittenAsEventSourceNotInConfiguredList(eventData.EventSource.Name);
+                }
+            }
+            catch (Exception ex)
+            {
+                EventCountersInstrumentationEventSource.Log.ErrorEventCounter(eventData.EventName, ex.ToString());
+            }
+        }
+
+        /// <inheritdoc/>
+        protected override void OnEventSourceCreated(EventSource eventSource)
+        {
+            // Keeping track of all EventSources here, as this call may happen before initialization.
+            this.allEventSourcesCreated.Enqueue(eventSource);
+
+            // options might be null when this method gets called
+            // before our constructor was completed
+            if (this.isInitialized)
+            {
+                this.EnableEventSource(eventSource);
```

Comment:
There's a small chance here that if we're not initialized that we loose the reference to the event source

Reply:
This scenario is already handled. The base EventListener constructor will call `OnEventSourceCreated` on every existing EventSource. Therefor our OnEventSourceCreated will [queue](https://github.com/hananiel/opentelemetry-dotnet-contrib/blob/b3a7d27320aaccf3a979b65f2fb445a60a29cd58/src/OpenTelemetry.Contrib.Instrumentation.EventCounters/Implementation/EventCounterListener.cs#L119) all eventsources and "[enable](https://github.com/hananiel/opentelemetry-dotnet-contrib/blob/b3a7d27320aaccf3a979b65f2fb445a60a29cd58/src/OpenTelemetry.Contrib.Instrumentation.EventCounters/Implementation/EventCounterListener.cs#L67)" then afterwards.

Reply:
Yup! This is well known and is handled by OpenTelemetry SDK as well for its internal logging (https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Internal/SelfDiagnosticsEventListener.cs)